java/ql/src/semmle/code/java/security/Random.qll

```diff
@@ -2,15 +2,183 @@ import java
 import semmle.code.java.dataflow.DefUse
 import semmle.code.java.dataflow.DataFlow
 
-class SecureRandomNumberGenerator extends RefType {
+/**
+ * A class with methods that generate random data.
+ */
+abstract class RandomNumberGenerator extends RefType { }
+
+/**
+ * The `java.security.SecureRandom` class.
+ */
+class SecureRandomNumberGenerator extends RandomNumberGenerator {
   SecureRandomNumberGenerator() { this.hasQualifiedName("java.security", "SecureRandom") }
 }
 
-class GetRandomData extends MethodAccess {
-  GetRandomData() {
-    this.getMethod().getName().matches("next%") and
-    this.getQualifier().getType() instanceof SecureRandomNumberGenerator
+/**
+ * The `java.util.Random` class or any of its subtypes, including `java.security.SecureRandom`.
+ */
+class StdlibRandom extends RandomNumberGenerator {
+  StdlibRandom() { this.getAnAncestor().hasQualifiedName("java.util", "Random") }
+}
+
+/**
+ * The `org.apache.commons.lang3.RandomUtils` class.
+ */
+class ApacheRandomUtils extends RandomNumberGenerator {
+  ApacheRandomUtils() { this.hasQualifiedName("org.apache.commons.lang3", "RandomUtils") }
+}
+
+/**
+ * A method access that returns random data or writes random data to an argument.
+ */
+abstract class RandomDataSource extends MethodAccess {
```

Comment:
The amount of virtual dispatch that goes on here is too convoluted for my tastes. And public abstract classes aren't good either. So we should think about how all of this could be rewritten.